Stronghold Improvements: Timeout and Unloading

jyhi opened the pull request:
This PR adds timeout related methods to `StrongholdAdapter`. `get_timeout()` and `set_timeout()` are added to view and change the key clearing timeout. The task will be automatically spawned in `set_timeout()` if a new timeout is provided and key is still in the memory. `restart_key_timeout()` is also added as a shortcut of refreshing the count down by setting the same timeout again.

Fix #996.

jyhi committed:
stronghold: add timeout and task methods

get_timeout() and set_timeout() are added to view and change the
key clearing timeout. The task will be automatically spawned in
set_timeout() if a new timeout is provided and key is still in the
memory.

restart_key_timeout() is also added as a shortcut of refreshing
the count down by setting the same timeout again.

jyhi committed:
secret: add timeout for dto ser/de

jyhi committed:
stronghold: no SecretManager methods without key

jyhi committed:
Revert "stronghold: no SecretManager methods without key"

This reverts commit 675ae1cd19017e3d29043e7f8b398a881d4d7e2a.

This check has no actual meaning - setting an incorrect key can
allow subsequent cryptography oprations run.

jyhi committed:
stronghold: kill stronghold on key clear

To further prevent cryptographic methods being invoked when the key
has been cleared, we have to unload Stronghold every time the key
is cleared from the memory, as otherwise Stronghold permits such
operation (it only asks for a key to read / write a snapshot).

There are also general code and documentation improvements.

jyhi committed:
stronghold: increase sleep duration in tests

Poor CI machines sometimes run out of time.

jyhi committed:
stronghold: unset snapshot load status on unload

jyhi committed:
secret: fix import warning

Thoralf-M commented on src/stronghold/mod.rs:
```diff
@@ -76,7 +78,7 @@ use crate::{Error, Result};
 pub struct StrongholdAdapter {
     /// A stronghold instance.
-    stronghold: Stronghold,
+    stronghold: Arc<Mutex<Stronghold>>,
```
Wouldn't a `RwLock` be better here?

Thoralf-M replied:
Probably doesn't really help

jyhi replied:
Maybe, but only `SecretManage` has 2 methods taking `&self`, while others take `&mut self`

Thoralf-M commented:
Shouldn't `clear_key()` also call `unload_stronghold_snapshot()`?

jyhi committed:
stronghold: unload on key clear

jyhi commented:
Oops, fixing.

Thoralf-M commented on src/stronghold/mod.rs:
```diff
@@ -280,6 +280,11 @@ impl StrongholdAdapter {
         }
+
+        // Unload Stronghold, but we can't do much about the errors.
+        if let Err(err) = self.unload_stronghold_snapshot().await {
```
This needs to be done before clearing the key, otherwise it will always fail

jyhi replied:
Oops, yes.

thibault-martinez commented on src/secret/mod.rs:
```diff
@@ -191,6 +195,7 @@ impl From<&SecretManager> for SecretManagerDto {
             #[cfg(feature = "stronghold")]
             SecretManager::Stronghold(stronghold_dto) => Self::Stronghold(StrongholdDto {
```
Not part of your PR but the variable shouldn't be called `*_dto`, would be nice if you could change it.

jyhi committed:
stronghold: fix stronghold unload on key clear

Unloading snapshot requires the key to flush the snapshot onto
disk, so the key must be retained upon unloading.

jyhi committed:
secret: fix variable namings

thibault-martinez commented on src/stronghold/mod.rs:
```diff
@@ -162,57 +168,61 @@ impl StrongholdAdapterBuilder {
 
+            // Stronghold isn't optional, so we must initialize one here if it hasn't been supplied.
+            let stronghold = self.stronghold.unwrap_or_else(|| {
+                // XXX: we unwrap here.
```
Weird comment?

jyhi replied:
It's for the next line; we're still on an older stable version of Stronghold, so we initialize `riker::ActorSystem` below, which returns a `Result`, but the `build()` method is not supposed to fail, so an `unwrap()` is left here.

It's actually a copy-pasta from the `Default` implementation (L118). To remove unwraps we can remove the `impl Default` and change `build() -> StrongholdAdapter` to `try_build() -> Result<StrongholdAdapter>`.

thibault-martinez replied:
Yeah but I mean, this comment is basically useless. We can see that we are unwrapping, we would rather be expecting an explanation of why we are 100% sure this is not going to panic. And replace XXX with `PANIC`.

jyhi replied:
Had a broader discussion and unwraps will be removed.

jyhi committed:
stronghold,secret: remove unwraps

Remove calls to .unwrap() to avoid panics. As a result, the build
method of StrongholdAdapterBuilder becomes failable and is renamed
to try_build. Existing codes are adjusted accordingly. General code
and documentation improvements are done too.
